### PR TITLE
feat: Models for Pokemon and Items with basic testing

### PR DESCRIPTION
**Pull request by mrestuccia**

### CODE REVIEW

File: server/models/migrations/20190617180248_create_pokemon.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import * as Knex from 'knex';
-import { PokeType } from '../PokeType';
+import { pokeType } from '../pokeType';
```
Comment (aldasmallthings) [inline server/models/migrations/20190617180248_create_pokemon.ts]: :) awesome nit catch

File: server/models/Pokemon.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+      },
+    },
+  };
+
+  // Custom Methods
+
+
+  // returns all Pokemon, ordered by pokemonNumber ascending
+  static async getAll(txn: Transaction): Promise<Pokemon[]> {
+    const pokemonList = await this.query(txn).whereNull('deletedAt');
```
Comment (aldasmallthings) [inline server/models/Pokemon.ts]: FYI :) you can order by pokemonNumber with `.orderBy('pokemonNumber');`

File: server/models/Pokemon.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    return pokemonList;
+  }
+
+  // get(pokemonId: string, txn: Transaction) ­ returns a single Pokemon, and associated items
+  // useful link: http://ivanbatic.com/using-async-await-typescript-classes/
+  static async get(pokemonId: string, txn: Transaction): Promise<Pokemon> {
+    const individualPokemon = await this.query(txn).findById(pokemonId).eager('item');
```
Comment (aldasmallthings) [inline server/models/Pokemon.ts]: you could get rid of items that don't have a deletedAt through:
`      .modifyEager('item', builder => builder.where('item.deletedAt', null))`

File: server/models/Pokemon.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+      imageUrl: { type: 'string', minLength: 1, maxLength: 255 },
+      createdAt: { type: 'timestamp' },
+      updatedAt: { type: 'timestamp' },
+      deletedAt: { type: 'timestamp' },
+    },
+  };
+
+
+  // Relations
+  static relationMappings: RelationMappings = {
```
Comment (aldasmallthings) [inline server/models/Pokemon.ts]: is this working as expected? I thought you needed a lemon `()` after relationMappings, as it's a static function... maybe i don't know the syntactical short hand, i'm new to using JS in the back end :)

File: server/models/_tests_/item.spec.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    const pokemonList = await Pokemon.getAll(trx);
+    const id = pokemonList[0].id;
+    const pokemon = await Pokemon.get(id, trx);
+    const itemId = pokemon.item[0].id;
+
+    // Query
+    const item = await Item.get(itemId, trx);
+
+    // Test
+    expect(item).toBeInstanceOf(Item);
```
Comment (aldasmallthings) [inline server/models/_tests_/item.spec.ts]: are you trying to test item's get method here and ensure it returns just one item?

File: server/models/_tests_/item.spec.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+      name: 'test item',
+      pokemonId: id,
+      price: 0,
+      happiness: 0,
+      imageUrl: 'test.png'
+    }, trx);
+
+    // Test
+    expect(createdItem).toBeInstanceOf(Item);
+    expect(createdItem.pokemonId).toBe(id);
```
Comment (aldasmallthings) [inline server/models/_tests_/item.spec.ts]: i think here you would want to check that the item you created has all of the same properties as createdItem :)

Comment (aldasmallthings) [reply server/models/_tests_/item.spec.ts]: you can use `toMatchObject` for that, if you want :)

File: server/models/_tests_/pokemon.spec.ts [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+
+  it('get: Returns one particular pokemon and related items', async () => {
+    const pokemonList = await Pokemon.getAll(trx);
+    const id = pokemonList[0].id
+    const pokemon = await Pokemon.get(id, trx)
+
+    // Test
+    expect(pokemon).toBeInstanceOf(Pokemon);
+    expect(pokemon).toHaveProperty('item');
```
Comment (aldasmallthings) [inline server/models/_tests_/pokemon.spec.ts]: if this were going into commons (not needed) i'd also check that all the right items were returned :)

File: server/models/Pokemon.ts (hunk 2 above)
Comment (mrestuccia) [reply server/models/Pokemon.ts]: Forgot to order it. :| Will fix it.

Comment (aldasmallthings) [reply server/models/Pokemon.ts]: 👨‍🎨

File: server/models/Pokemon.ts (hunk 3 above)
Comment (mrestuccia) [reply server/models/Pokemon.ts]: I didn't even consider to not display the items that were deleted. Thanks!

File: server/models/Pokemon.ts (hunk 4 above)
Comment (mrestuccia) [reply server/models/Pokemon.ts]: I think it's correct syntactically if you use typeScript.

See: https://github.com/Vincit/objection.js/blob/master/examples/express-ts/src/models/Animal.ts

Also it's working as I pull the relationships the proper way.

File: server/models/_tests_/item.spec.ts (hunk 5 above)
Comment (mrestuccia) [reply server/models/_tests_/item.spec.ts]: On. I added that should not be an array. I think that is suffice. I am casting that is item type and not an array.

File: server/models/_tests_/item.spec.ts (hunk 6 above)
Comment (mrestuccia) [reply server/models/_tests_/item.spec.ts]: This is excellent. Fixed. Thanks!

File: server/models/_tests_/item.spec.ts [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    expect(item).toBeInstanceOf(Item);
+    expect(item).not.toBeInstanceOf(Array);
+
+  });
+
+
+  it('create: Creates and returns an item', async () => {
+
+    // Get a pokemon id
+    const pokemonList = await Pokemon.getAll(trx);
```
Comment (foxdoubt) [inline server/models/_tests_/item.spec.ts]: I think a more stable version of this test would be as follows:

```javascript
// step 1: create a Pokemon
const createdItem = await Item.create(testItem, trx);
// step 2: check to see if it's in the database
const fetchedFromDb = await Item.query(txn).findById(createdItem.id, trx)
expect(createdItem.id).toEqual(fetchedFromDb.id);
```

File: server/models/_tests_/item.spec.ts [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+    const id = pokemonList[0].id;
+    const pokemon = await Pokemon.get(id, trx);
+    const itemId = pokemon.item[0].id;
+
+    // Delete Pokemon
+    const deletedItem = await Item.delete(itemId, trx);
+
+    // Test
+    expect(deletedItem).toBeInstanceOf(Item);
```
Comment (foxdoubt) [inline server/models/_tests_/item.spec.ts]: Here, I think a more exact test would be, after deleting the item, calling `Item.getAll` and testing to make sure your deleted item is not included in the query results. 

```javaScript
const allItems = await Item.getAll(trx);
expect(allItems).not.toEqual(expect.arrayContaining([expect.objectContaining({id: deletedItem.id})])
```
Or something to that effect, where you check that the item is _not_ in the result.

File: server/models/_tests_/item.spec.ts (hunk 8 above)
Comment (mrestuccia) [reply server/models/_tests_/item.spec.ts]: OK. Cool. Changed as suggested.

File: server/models/_tests_/item.spec.ts (hunk 9 above)
Comment (mrestuccia) [reply server/models/_tests_/item.spec.ts]: OK. I don't have a Item.getAll, but I retreived the same pokemon and checked that the item is no longer there.
This make sense. Thanks!

**Pull request merged**